File: 2025/day7/main.py
from dataclasses import dataclass
from collections import deque


@dataclass
class Grid:
    start: tuple[int]
    grid: list[list]
    rows: int
    cols: int


def read_file(fname):
    with open(fname) as f:
        lines = [line.rstrip("\n") for line in f.readlines() if line.strip()]
    return lines


def parser(data):
    rows = len(data)
    cols = len(data[0])
    start = (-1, -1)

    grid = [['.' for _ in range(cols)] for _ in range(rows)]

    for i in range(rows):
        for j in range(cols):
            curr = data[i][j]
            grid[i][j] = curr
            if curr == "S":
                start = (i, j)

    return Grid(start=start, grid=grid, rows=rows, cols=cols)


def bfs(grid: Grid):
    queue = deque([grid.start])
    splits = 0
    visited = set()

    while queue:
        row, col = queue.popleft()

        if row < 0 or row > grid.rows - 1 or col < 0 or col > grid.cols - 1:
            continue

        if (row, col) in visited:
            continue

        visited.add((row, col))

        cell = grid.grid[row][col]

        # reached the bottom
        if row == grid.rows - 1:
            continue

        if cell != "^":
            # go down
            queue.append((row + 1, col))
        else:
            # split beam, go left and right
            splits += 1
            queue.append((row, col - 1))
            queue.append((row, col + 1))

    return splits


def dfs(grid: Grid):
    memo = dict()

    def _helper(row, col, visited):
        if row < 0 or row > grid.rows - 1 or col < 0 or col > grid.cols - 1:
            return 0

        if (row, col) in visited:
            return 0

        if row == grid.rows - 1:
            return 1

        if (row, col) not in visited and (row, col) in memo:
            return memo[(row, col)]

        cell = grid.grid[row][col]
        path_visited = visited.union({(row, col)})

        if cell == "^":
            left = _helper(row + 1, col - 1, path_visited)
            right = _helper(row + 1, col + 1, path_visited)
            result = left + right
        else:
            result = _helper(row + 1, col, path_visited)

        if (row, col) not in visited:
            memo[(row, col)] = result

        return result

    return _helper(grid.start[0], grid.start[1], set())


def solve(fname, part1=True):
    lines = read_file(fname)
    grid = parser(lines)
    # print(grid)
    if part1:
        tachyon_splits = bfs(grid)
        total = tachyon_splits
    else:
        timelines = dfs(grid)
        total = timelines
    return total


def part1():
    # result = solve("sample.txt")
    result = solve("input.txt")
    print(f"Part 1: {result}")


def part2():
    # result = solve("sample.txt", False)
    result = solve("input.txt", False)
    print(f"Part 2: {result}")


if __name__ == "__main__":
    part1()
    part2()

File: 2025/day7/src/main.rs
use aoc_utils::{FromGrid, parse_grid_from};
use std::collections::{HashMap, HashSet, VecDeque};
use std::rc::Rc;

type Coord = (usize, usize);

#[derive(Debug, Clone, PartialEq, Eq)]
struct Point {
    x: usize,
    y: usize,
    value: char,
}

impl Point {
    fn new(x: usize, y: usize, value: char) -> Self {
        Self { x, y, value }
    }

    fn coords(&self) -> Coord {
        (self.x, self.y)
    }

    fn is_start(&self) -> bool {
        self.value == 'S'
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct Grid {
    start: Coord,
    data: Vec<Vec<Point>>,
}

impl Grid {
    fn new(data: Vec<Vec<Point>>) -> Self {
        let mut start = None;

        for row in &data {
            for point in row {
                if point.is_start() {
                    start = Some(point.coords());
                    break;
                }
            }
            if start.is_some() {
                break;
            }
        }

        let start = start.expect("No starting point found");
        Self { start, data }
    }

    fn at(&self, row: usize, col: usize) -> char {
        self.data[row][col].value
    }

    fn start(&self) -> Coord {
        self.start
    }

    fn num_rows(&self) -> usize {
        self.data.len()
    }

    fn num_cols(&self) -> usize {
        self.data[0].len()
    }

    fn in_bounds(&self, row: usize, col: usize) -> bool {
        row < self.num_rows() && col < self.num_cols()
    }
}

impl FromGrid for Grid {
    type Element = Point;

    fn parse_element(row: usize, col: usize, c: char) -> Self::Element {
        Point::new(row, col, c)
    }

    fn from_grid(grid: Vec<Vec<Self::Element>>) -> Self {
        Grid::new(grid)
    }
}

fn bfs(grid: &Grid) -> u64 {
    let mut splits = 0;
    let mut queue: VecDeque<Coord> = VecDeque::new();
    let mut visited: HashSet<Coord> = HashSet::new();

    queue.push_back(grid.start());

    while let Some((row, col)) = queue.pop_front() {
        if !grid.in_bounds(row, col) {
            continue;
        }

        if visited.contains(&(row, col)) {
            continue;
        }
        visited.insert((row, col));

        if row == grid.num_rows() - 1 {
            continue;
        }

        let cell = grid.at(row, col);

        if cell != '^' {
            queue.push_back((row + 1, col));
        } else {
            splits += 1;
            if col > 0 {
                queue.push_back((row, col - 1));
            }
            queue.push_back((row, col + 1));
        }
    }

    splits
}

fn dfs(grid: &Grid) -> usize {
    let mut memo: HashMap<Coord, usize> = HashMap::new();

    fn _recurse(
        grid: &Grid,
        row: usize,
        col: usize,
        visited: Rc<HashSet<Coord>>,
        memo: &mut HashMap<Coord, usize>,
    ) -> usize {
        if !grid.in_bounds(row, col) {
            return 0;
        }
        if visited.contains(&(row, col)) {
            return 0;
        }

        if row == grid.num_rows() - 1 {
            return 1;
        }

        if let Some(&cached) = memo.get(&(row, col)) {
            return cached;
        }

        let mut path_visited = (*visited).clone();
        path_visited.insert((row, col));
        let path_visited = Rc::new(path_visited);

        let cell = grid.at(row, col);

        let result = if cell == '^' {
            let left = if col > 0 {
                _recurse(grid, row + 1, col - 1, Rc::clone(&path_visited), memo)
            } else {
                0
            };
            let right = _recurse(grid, row + 1, col + 1, Rc::clone(&path_visited), memo);
            left + right
        } else {
            _recurse(grid, row + 1, col, Rc::clone(&path_visited), memo)
        };

        if !visited.contains(&(row, col)) {
            memo.insert((row, col), result);
        }

        result
    }

    let (x, y) = grid.start();
    _recurse(grid, x, y, Rc::new(HashSet::new()), &mut memo)
}

fn part1(grid: &Grid) -> u64 {
    bfs(grid)
}

fn part2(grid: &Grid) -> usize {
    dfs(grid)
}

fn main() -> std::io::Result<()> {
    let grid: Grid = parse_grid_from("2025/day7/input.txt")?;

    println!("Part 1: {}", part1(&grid));
    println!("Part 2: {}", part2(&grid));

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_INPUT_FULL: &str = "\
.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............";

    fn parse_test_input() -> Grid {
        Grid::from_grid(
            TEST_INPUT_FULL
                .lines()
                .enumerate()
                .map(|(row_index, line)| {
                    line.trim()
                        .chars()
                        .enumerate()
                        .map(|(col_index, c)| Grid::parse_element(row_index, col_index, c))
                        .collect()
                })
                .collect(),
        )
    }

    #[test]
    fn test_parse() {
        let grid = parse_test_input();
        assert_eq!(grid.num_rows(), 16);
        assert_eq!(grid.num_cols(), 15);
        assert_eq!(grid.start, (0, 7));
    }

    #[test]
    fn test_part1() {
        let grid = parse_test_input();
        assert_eq!(part1(&grid), 21);
    }

    #[test]
    fn test_part2() {
        let grid = parse_test_input();
        assert_eq!(part2(&grid), 40);
    }
}
